pmda/density.py

```diff
@@ -289,20 +289,31 @@ def _prepare(self):
         grid, edges = np.histogramdd(np.zeros((1, 3)), bins=bins,
                                      range=arange, normed=False)
         grid *= 0.0
-        self._grid = grid
+
+        self._results = [grid] * self.n_frames
         self._edges = edges
         self._arange = arange
         self._bins = bins
 
-    def _single_frame(self, ts, atomgroups):
-        coord = self.current_coordinates(atomgroups[0], self._atomselection,
-                                         self._updating)
-        result = np.histogramdd(coord, bins=self._bins, range=self._arange,
-                                normed=False)
-        return result[0]
+    def _single_frame(self):
+        h, _ = np.histogramdd(self._atomgroup.positions,
+                              bins=self._bins, range=self._arange,
+                              normed=False)
+        # reduce (proposed change #2542 to match the parallel version in pmda.density)
+        # return self._reduce(self._grid, h)
+        #
+        # serial code can simply do
+
+        # the current timestep of the trajectory is self._ts
+#        self._results[self._frame_index][0] = self._ts.frame
+        # the actual trajectory is at self._trajectory
+#        self._results[self._frame_index][1] = self._trajectory.time
+        self._results[self._frame_index] = h
```

Comment:
Storing a full histogram for each frame is bad – you can easily run out of memory. I think it is important that the aggregation is done every step and not just in `_conclude`.

Reply:
But isn't that what also happens with `_reduce`? It won't pass the full histogram back to the main process, but only the calculated frames in `_dask_helper`.

Reply:
No, the density reduce https://github.com/MDAnalysis/pmda/blob/13fa3b5cf0055b8b4c3b45056bc740659815c624/pmda/density.py#L326-L332 does an in-place _addition_ (no a list append) in line 331. In _conclude https://github.com/MDAnalysis/pmda/blob/13fa3b5cf0055b8b4c3b45056bc740659815c624/pmda/density.py#L305-L306 we only sum over the summed histograms of the blocks and then divide by all frames to get the average.

Reply:
Btw, the PMDA paper has a discussion on that topic.